spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/lock/ZookeeperLockRegistry.java

```diff
@@ -0,0 +1,261 @@
+		Assert.isInstanceOf(String.class, lockKey);
+		String path = this.keyToPath.pathFor((String) lockKey);
+		ZkLock lock = this.locks.get(path);
+		synchronized (this.locks) {
+			lock = this.locks.get(path);
+			if (lock == null) {
+				lock = new ZkLock(this.client, path);
+				this.locks.put(path, lock);
+			}
+			if (this.trackingTime) {
+				lock.setLastUsed(System.currentTimeMillis());
+			}
+		}
+		return lock;
+
+	}
+
+	/**
+	 * Remove locks last acquired more than 'age' ago that are not currently locked.
+	 * Expiry is not supported if the {@link KeyToPathStrategy} is bounded (returns a finite
+	 * number of paths). With such a {@link KeyToPathStrategy}, the overhead of tracking when
+	 * a lock is obtained is avoided.
+	 * @param age the time since the lock was last obtained.
+	 */
+	@Override
+	public void expireUnusedOlderThan(long age) {
+		if (!this.trackingTime) {
+			throw new IllegalStateException("Ths KeyToPathStrategy is bounded; expiry is not supported");
+		}
+		synchronized(this.locks) {
+			Iterator<Entry<String, ZkLock>> iterator = this.locks.entrySet().iterator();
+			long now = System.currentTimeMillis();
+			while(iterator.hasNext()) {
+				Entry<String, ZkLock> entry = iterator.next();
+				ZkLock lock = entry.getValue();
+				if (now - lock.getLastUsed() > age
+						&& !lock.isAcquiredInThisProcess()) {
+					iterator.remove();
+				}
+			}
+		}
+	}
+
+	/**
+	 * Strategy to convert a lock key (e.g. aggregation correlation id) to a
+	 * Zookeeper path.
+	 *
+	 */
+	public interface KeyToPathStrategy {
+
+		/**
+		 * Return the path for the key.
+		 * @param key the key.
+		 * @return the path.
+		 */
+		String pathFor(String key);
+
+		/**
+		 * @return true if this strategy returns a bounded number of locks, removing
+		 * the need for removing LRU locks.
+		 */
+		boolean bounded();
+
+	}
+
+	private static class DefaultKeyToPathStrategy implements KeyToPathStrategy {
+
+		private final String root;
+
+		public DefaultKeyToPathStrategy(String rootPath) {
+			Assert.notNull(rootPath, "'rootPath' cannot be null");
+			if (!rootPath.endsWith("/")) {
+				this.root = rootPath + "/";
+			}
+			else {
+				this.root = rootPath;
+			}
+		}
+
+		@Override
+		public String pathFor(String key) {
+			return root + key;
+		}
+
+		@Override
+		public boolean bounded() {
+			return false;
+		}
+
+	}
+
+	private static class ZkLock implements Lock {
+
+		private final InterProcessMutex mutex;
+
+		private final String path;
+
+		private long lastUsed;
+
+		public ZkLock(CuratorFramework client, String path) {
+			this.mutex = new InterProcessMutex(client, path);
+			this.path = path;
+		}
+
+		public long getLastUsed() {
+			return this.lastUsed;
+		}
+
+		public void setLastUsed(long lastUsed) {
+			this.lastUsed = lastUsed;
+		}
+
+		@Override
+		public void lock() {
+			try {
+				this.mutex.acquire();
+			}
+			catch (Exception e) {
+				throw new MessagingException("Failed to aquire mutex at " + this.path, e);
```

Comment:
This isn't a Messaging component and from other side using this one within a Messaging component, raising a `MessagingException` from here, we prevent re-throwing a correct `MessagingException` with `failedMessage` context.
So, for me simple `RuntimeException` is enough in this context.
WDYT?